public/app/features/trails/utils.ts

```diff
@@ -115,3 +118,42 @@ export function getFilters(scene: SceneObject) {
   }
   return null;
 }
+
+// frontend hardening limit
+const MAX_ADHOC_VARIABLE_OPTIONS = 10000;
+
+export function limitAdhocProviders(
```

Comment:
Can we have unit test for this function?

Reply:
Added unit tests to check the function!